gateway/src/apicast/policy/keycloak_role_check/keycloak_role_check.lua

```diff
@@ -0,0 +1,195 @@
+
+local new = _M.new
+
+local function build_templates(scopes)
+  for _, scope in ipairs(scopes) do
+
+    if scope.realm_roles then
+      for _, realm_role in ipairs(scope.realm_roles) do
+        realm_role.template_string = TemplateString.new(
+          realm_role.name, realm_role.name_type or default_type)
+      end
+    end
+
+    if scope.client_roles then
+      for _, client_role in ipairs(scope.client_roles) do
+        client_role.name_template_string = TemplateString.new(
+          client_role.name, client_role.name_type or default_type)
+
+        client_role.client_template_string = TemplateString.new(
+          client_role.client, client_role.client_type or default_type)
+      end
+    end
+
+    scope.resource_template_string = TemplateString.new(
+      scope.resource, scope.resource_type or default_type)
+
+  end
+end
+
+function _M.new(config)
+  local self = new()
+  self.config = config or {}
+  self.type = config.type or "whitelist"
+  self.scopes = config.scopes or {}
+
+  build_templates(self.scopes)
+
+  return self
+end
+
+local function check_roles_in_token(role, roles_in_token)
+  for _, role_in_token in ipairs(roles_in_token) do
+    if role == role_in_token then return true end
+  end
+
+  return false
+end
+
+local function match_realm_roles(scope, context)
+  if not scope.realm_roles then return true end
+
+  for _, role in ipairs(scope.realm_roles) do
+    if not context.jwt.realm_access then
+      return false
+    end
+
+    local name = role.template_string:render(context)
+
+    if not check_roles_in_token(name, context.jwt.realm_access.roles) then
+      return false
+    end
+  end
+
+  return true
+end
+
+local function match_client_roles(scope, context)
+  if not scope.client_roles then return true end
+
+  for _, role in ipairs(scope.client_roles) do
+    if not context.jwt.resource_access then
+      return false
+    end
+
+    local client = role.client_template_string:render(context)
+    local client_in_token = context.jwt.resource_access[client]
+
+    if not client_in_token then
+      ngx.log(ngx.DEBUG, "Client '", client, "' was not found in the access token.")
+      return false
+    end
+
+    local name = role.name_template_string:render(context)
+
+    if not check_roles_in_token(name, client_in_token.roles) then
+      return false
+    end
+  end
+
+  return true
+end
+
+local function scope_check(scopes, context)
+  local uri = ngx.var.uri
+
+  for _, scope in ipairs(scopes) do
+
+    local resource = scope.resource_template_string:render(context)
+
+    local mapping_rule = MappingRule.from_proxy_rule({
+      http_method = 'ANY',
+      pattern = resource,
+      querystring_parameters = {},
+      metric_system_name = 'hits'
+    })
+
+    if mapping_rule:matches('ANY', uri) then
+      if match_realm_roles(scope, context) and match_client_roles(scope, context) then
+        return true
+      end
+    end
+
+  end
+
+  return false
+end
+
+local function auth_failed(service)
```

Comment:
We should probably reuse the method that we have in the `Errors` module instead:
https://github.com/3scale/apicast/blob/a95ebddd2bc25e953fb8096ce181cd5123ec20dd/gateway/src/apicast/errors.lua#L12

Reply:
Changed.